New Analyzer: Warn when calling Enumerable.Cast<T>/OfType<T> with incompatible types

fowl2 opened the pull request:
Warn if use of `Enumerable.Cast<T>` would always fail at runtime or `OfType<T>()` would provably result in an empty sequence, because we can see that the input type in the sequence will never be the specified type.

RuleId: CA2019
Category: Reliability

related: https://github.com/dotnet/roslyn-analyzers/issues/3608, https://github.com/dotnet/runtime/issues/33770

- [x] Tests pass
- [x] Targeted to ~release/7.0.1xx~ main
- [x] Run against 
   - [x] dotnet/roslyn-analyzers 
   - [x] dotnet/runtime
   - [ ] dotnet/roslyn 
- [ ] Docs

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=4328) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=4328) before we can accept your contribution.<br/><br/>:x: fowl2 [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn-analyzers?pullRequest=4328)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn-analyzers?pullRequest=4328) it.</sub>

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+
+                    var targetMethod = invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod;
+                    if (targetMethod == null || !methodMetadataNames.TryGetValue(targetMethod.Name, out var rule))
```
TargetMethod cannot be null
```suggestion
                    if (!methodMetadataNames.TryGetValue(targetMethod.Name, out var rule))
```

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+                                                                             s_localizableOfTypeMessage,
+                                                                             DiagnosticCategory.Reliability,
+                                                                             RuleLevel.IdeSuggestion,
+                                                                             description: null,
```
We generally recommend specifying a description with detailed explanation of the issue and recommended fix.

fowl2 commented:
That should be all the warnings resolved except for the rule id. Not sure how those are are assigned.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1089,6 +1089,24 @@
+    <value>If the source sequence contains any elements, the sequence returned by Cast will throw InvalidCastException at runtime when enumerated. Change the type(s) to those that are compatible or remove the call.</value>
+  </data>
+  <data name="DoNotCallEnumerableCastThatWillFailMessage" xml:space="preserve">
+    <value>Enumerable.Cast&lt;T&gt; will '{0}' can not be castetype '{1}' will fail</value>
```
What about `Type '{0}' cannot be converted to '{1}' and will lead to a runtime failure`?

fowl2 replied:
Is 'converted' the correct terminology?  How about `Type '{0}' is incompatible with type '{1}' and will cause an exception at runtime`?

Evangelink replied:
From the doc page they say:

> If an element cannot be converted to type `TResult`, this method throws a `InvalidCastException`.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1089,6 +1089,24 @@
     <value>When creating path for '{0} in method {1}' from relative archive item path to extract file and the source is an untrusted zip archive, make sure to sanitize relative archive item path '{2} in method {3}'</value>
   </data>
+  <data name="DoNotCallEnumerableCastThatWillFailDescription" xml:space="preserve">
+    <value>If the source sequence contains any elements, the sequence returned by Cast will throw InvalidCastException at runtime when enumerated. Change the type(s) to those that are compatible or remove the call.</value>
```
```suggestion
    <value>If the source sequence contains any element, the call to 'Cast' will throw an  'InvalidCastException' at runtime when enumerated. Make sure the types are compatible or remove this call.</value>
```

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1089,6 +1089,24 @@
+    <value>Do not call Enumerable.Cast&lt;T&gt; with incompatible types</value>
+  </data>
+  <data name="DoNotCallEnumerablOfTypeThatWillNeverReturnAnyValuesMessage" xml:space="preserve">
+    <value>Source sequence elements are of type '{1}' which is not compatible with provided result element type '{0}'</value>
```
`Type '{0}' cannot be converted to '{1}' and will results in empty sequence`

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1089,6 +1089,24 @@
+    <value>The enumerable returned by this call to OfType&lt;T&gt; will never return any values</value>
+  </data>
+  <data name="DoNotCallEnumerablOfTypeThatWillNeverReturnAnyValuesMessageDescription" xml:space="preserve">
+    <value>The type of the elements in the source sequence is not compatible with the type passed to OfType&lt;T&gt;, therefore no elements will ever be returned. Change the type of the elements in the source sequence, the type passed to OfType&lt;T&gt;, or remove the call.</value>
```
```suggestion
    <value>This call to 'OfType' will always results in an empty sequence. Make sure the types are compatible or remove this call.</value>
```

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotCallEnumerableCastThatWillFailAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA9999";
```
Let's use the next available in `Usage` (i.e. `CA2250`), see https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/usage-warnings

Evangelink replied:
Except if I am missing something this isn't fixed.

fowl2 replied:
Sorry, I'm ticking them off as I go. Pushed now fix now.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+                                                                             s_localizableCastTitle,
+                                                                             s_localizableCastMessage,
+                                                                             DiagnosticCategory.Reliability,
+                                                                             RuleLevel.BuildWarningCandidate,
```
Use `IdeSuggestion`. A rule should never be promoted to warning without having been verified and validated by MS.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+                                                                             s_localizableOfTypeTitle,
+                                                                             s_localizableOfTypeMessage,
+                                                                             DiagnosticCategory.Reliability,
+                                                                             RuleLevel.BuildWarningCandidate,
```
Same remark.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var enumerableType = compilationStartContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemLinqEnumerable);
```
You could use the `TryGet` method as you bail-out when any is not present.

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(compilationStartContext =>
```
Use `context` instead of `compilationStartContext`

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+                    return;
+                }
+
+                compilationStartContext.RegisterOperationAction(operationContext =>
```
Use `context` instead of `operationContext`

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+
+                    var targetMethod = invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod;
```
```suggestion
                    var targetMethod = invocation.TargetMethod.OriginalDefinition;
```

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+
+                    var targetMethod = invocation.TargetMethod.ReducedFrom ?? invocation.TargetMethod;
+                    if (!methodMetadataNames.TryGetValue(targetMethod.Name, out var rule))
```
We usually prefer a symbol comparison rather than a name comparison.

fowl2 replied:
How the the approach I've just pushed?

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+                        return;
+                    }
+
+                    var arg = invocation.IsExtensionMethodAndHasNoInstance()
```
I need to look for it but I am sure that we have something similar and that there is a little trick for vbnet.

fowl2 replied:
this does seem pretty messy. I remember looking for quite a while but couldn't find anything.

fowl2 replied:
found it! `GetInstance()`

Evangelink commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzerTests.cs:
```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+
```
```suggestion
```

Evangelink commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzerTests.cs:
```diff
@@ -0,0 +1,130 @@
+using System.Threading.Tasks;
+
+using Microsoft.CodeAnalysis.Testing;
+
```
```suggestion
```

Evangelink commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzerTests.cs:
```diff
@@ -0,0 +1,130 @@
+using Microsoft.CodeAnalysis.Testing;
+
+using Xunit;
+
```
```suggestion
```

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,171 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
```
```suggestion
```

Evangelink commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzerTests.cs:
```diff
@@ -0,0 +1,130 @@
+}
+",
+    // /0/Test0.cs(9,17): info CA9999: If the source sequence contains any elements, the sequence returned by Cast will throw InvalidCastException at runtime when enumerated.
+    VerifyCS.Diagnostic(ofTypeRule).WithSpan(9, 17, 9, 46).WithArguments("int", "string")
```
Could you please use inline markup notation rather than `WithSpan` it is easier to read and comprehend (e.g. `var a = (new int[0]).{|#0:OfType<object>()|}`

fowl2 replied:
This is the format output by the test runner - is there any way to get it in this format or will this have to be done manually?

fowl2 replied:
Will do.

fowl2 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1089,6 +1089,24 @@
+    <value>Enumerable.Cast&lt;T&gt; will '{0}' can not be castetype '{1}' will fail</value>
+  </data>
+  <data name="DoNotCallEnumerableCastThatWillFailTitle" xml:space="preserve">
+    <value>Do not call Enumerable.Cast&lt;T&gt; with incompatible types</value>
```
I don't think rule titles support format items, I can't see any others that are using them.

fowl2 replied:
ok I'll leave it as

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1,17 +1,17 @@
     Version 2.0
-
-    The primary goals of this format is to allow a simple XML format
-    that is mostly human readable. The generation and parsing of the
-    various data types are done through the TypeConverter classes
+    
+    The primary goals of this format is to allow a simple XML format 
+    that is mostly human readable. The generation and parsing of the 
+    various data types are done through the TypeConverter classes 
     associated with the data types.
-
+    
     Example:
-
+    
```
Was this file edited manually?
It's strange to see these extra spaces.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1089,6 +1089,24 @@
     <value>When creating path for '{0} in method {1}' from relative archive item path to extract file and the source is an untrusted zip archive, make sure to sanitize relative archive item path '{2} in method {3}'</value>
   </data>
+  <data name="DoNotCallEnumerableCastThatWillFailDescription" xml:space="preserve">
+    <value>If the source sequence contains any element, the call to 'Cast' will throw an  'InvalidCastException' at runtime when enumerated. Make sure the types are compatible or remove this call.</value>
```
remove extra space.

```suggestion
    <value>If the source sequence contains any element, the call to 'Cast' will throw an 'InvalidCastException' at runtime when enumerated. Make sure the types are compatible or remove this call.</value>
```

Evangelink commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,169 @@
+                    case (TypeKind.Struct, TypeKind.Struct):
+                    case (TypeKind.Struct, TypeKind.Enum):
+                    case (TypeKind.Enum, TypeKind.Struct):
+                        if (castFrom.DerivesFrom(castTo))
+                        {
+                            return false;
+                        }
+
+                        if (castTo.DerivesFrom(castFrom))
+                        {
+                            return false;
+                        }
+
+                        return true;
```
```suggestion
                        return !castFrom.DerivesFrom(castTo) &&
                        	!castTo.DerivesFrom(castFrom));
```

Evangelink commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzerTests.cs:
```diff
@@ -0,0 +1,127 @@
+        }
+
+        [Fact]
+        public async Task DiagnosticCasesVB()
```
It'd be nice to have an equivalent in term of cases used for C# and vbnet.

fowl2 replied:
I'll (hopefully) get a chance to work on improving the tests this weekend. @Evangelink if you could provide feedback on the other comments, that's be great.

fowl2 replied:
I added a bunch of c# tests, not exactly equivalent, but hopefully enough.

fowl2 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -1089,6 +1089,24 @@
+    <value>Type '{0}' is incompatible with type '{1}' and will result in an empty sequence</value>
+  </data>
+  <data name="DoNotCallEnumerablOfTypeThatWillNeverReturnAnyValuesTitle" xml:space="preserve">
+    <value>The enumerable returned by this call to OfType&lt;T&gt; will never return any values</value>
```
the other strings don't seem to quote these

fowl2 replied:
ahh I'll do it anyway

fowl2 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,169 @@
+                        return;
+                    }
+
+                    var argIEnumerableType = conversionOperation.Operand.Type.AllInterfaces.SingleOrDefault(t => t.OriginalDefinition.SpecialType == SpecialType.System_Collections_Generic_IEnumerable_T);
```
I have no idea what more than one `System.Collections.Generic.IEnumerable<T>` would means. I think a failure would be justified?

mavasani replied:
Yes, we never use `SingleOrDefault` in analyzers looking for specific symbol structure as that can cause analyzer crash for invalid code. I agree on using `FirstOrDefault`. Additionally, `Operand.Type` can be null as well, so needs to be `Operand.Type?.AllInterfaces`

buyaa-n reviewed (changes requested):
@fowl2 The analyzer proposal has not yet been approved, please wait until the proposal goes through API review

carlossanlop commented:
The proposal has been approved so we can unblock the PR, @buyaa-n

fowl2 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,169 @@
+                var methodRuleDictionary = methodMetadataNames
+                    .SelectMany(m => enumerableType.GetMembers(m.MethodName)
+                                                   .OfType<IMethodSymbol>()
+                                                   .Where(method => method.TypeParameters.HasExactly(1)
+                                                                 && method.Parameters.HasExactly(1))
```
@mavasani any thoughts on the best way of locating a particular member?

fowl2 replied:
Am I understanding that would cause this not to trigger without the extension method syntax?

ie. is there a reason why there should be a difference between
 `(new string[] { "not an int" }).Cast<int>();` and
 `Enumerable.Cast<int>(new string[] { "not an int" });`?

buyaa-n replied:
> Am I understanding that would cause this not to trigger without the extension method syntax?

Yes, and it is better to only restrict to the existing implementation of `Enumerable.Cast<T>/OfType<T>` which takes `this` parameter. There is no other overload for Enumerable.Cast<T>,  and if an overload is added in the future we don't know that overload would have the same behavior and needs to warn

buyaa-n replied:
> ie. is there a reason why there should be a difference between

I don't see any difference between them for the analyzer

fowl2 replied:
ahh, I was confusing method definition and call site. I used `method.IsExtensionMethod` instead.

fowl2 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,169 @@
+
+                    if (CastWillAlwaysFail(castFrom, castTo))
+                    {
+                        context.ReportDiagnostic(invocation.CreateDiagnostic(rule, castFrom.ToDisplayString(), castTo.ToDisplayString()));
```
@mavasani is `ToDisplayString` ok in a diagnostic? Some existing analyzers do use it.

buyaa-n replied:
It should be OK, we might prefer to set the display format though

mavasani replied:
`ToDisplayString` can produce really large strings in general, but I think it would be fine for type arguments.

buyaa-n commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzerTests.cs:
```diff
@@ -0,0 +1,127 @@
+    {
+        var a = (new int[0]).OfType<object>();
+        var b = (new int[0]).OfType<string>();
+        var c = (new object[0]).OfType<string>();
```
Could we have tests for `.Cast<T>()`?

MaStr11 replied:
Would also be good to have a test for `.Cast<T>` in [query expression syntax](https://docs.microsoft.com/en-US/dotnet/csharp/programming-guide/concepts/linq/converting-data-types)

```CS
_ = from string s in new object[0]
    select s;
```

fowl2 replied:
added some for both

buyaa-n reviewed (changes requested):
Overall looks good @fowl2 thank you for your contribution, added few comment, and seems we could add some more unit tests.
Not sure if you read our [Definition of done](https://github.com/dotnet/roslyn-analyzers/blob/master/docs/NetCore_GettingStarted.md#definition-of-done) list, it would be great if you could try testing the new analyzer with runtime repo, please let us know if you would like to give it a try

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,160 @@
+                        return;
+                    }
+
+                    var castFrom = argIEnumerableType.TypeArguments[0];
```
Can the type argument count be 0 for invalid code? I'd rather be defensive and check `argIEnumerableType.TypeArguments.IsEmpty` in the above bail out check.

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,160 @@
+                    }
+
+                    var castFrom = argIEnumerableType.TypeArguments[0];
+                    var castTo = invocation.TargetMethod.TypeArguments[0];
```
Same here - we should check for `invocation.TargetMethod.TypeArguments.IsEmpty`

mavasani commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,160 @@
+                }, OperationKind.Invocation);
+            });
+
+            static bool CastWillAlwaysFail(ITypeSymbol castFrom, ITypeSymbol castTo)
```
Hmm, I am wondering if we can use any compiler API here instead of trying to mimic conversion logic. @333fred Is there any compiler API that given two type symbols returns the `Conversion` info for it?

fowl2 replied:
One little/important detail here is that there are some subtle differences between the language  and IL casts. Eg. C# will emit calls to explicit and numeric conversion operators but the runtime will not. 

Note to self: ensure there is a test for this.

fowl2 replied:
I couldn't end up finding any API for this.

mavasani commented:
Please retarget this PR to https://github.com/dotnet/roslyn-analyzers/tree/release/6.0.1xx-preview1

mavasani reviewed (dismissed):
Please retarget this PR to https://github.com/dotnet/roslyn-analyzers/tree/release/6.0.1xx-preview1

buyaa-n commented:
@fowl2 could you address the comments? Do you need a help with anything?

Also please retarget to https://github.com/dotnet/roslyn-analyzers/tree/release/6.0.1xx

fowl2 commented:
@buyaa-n sorry I haven't had time to look at this - looks like I'll have some spare time this week to get back to you

danmoseley commented:
@fowl2 could you give us an update? It would be great to have this analyzer. Do you expect to have more time?

buyaa-n commented:
ping @fowl2 again, are you planning to work on it

fowl2 commented:
@buyaa-n it looks like the tooling doesn't really like having two different messages for the same rule id, should I split this into `CA2017` and `CA2018`?

buyaa-n commented:
> @buyaa-n it looks like the tooling doesn't really like having two different messages for the same rule id, should I split this into CA2017 and CA2018?

Why? Different messages should be OK, it seems some more files not generated yet, please run `msbuild /t:pack`

I see the message : `One or more auto-generated documentation files were either edited manually, or not updated. Please revert changes made to the following files (if manually edited) and run "msbuild /t:pack" for each solution at the root of the repo to automatically update them` in build failure

fowl2 commented:
> Why? Different messages should be OK, it seems some more files not generated yet, please run `msbuild /t:pack`

After doing this `src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.{md,sarif}` only contains the first description.
 
> By the way you don't need to retarget to `6.0.0-xyz` branches anymore as 6.0.0 is cut for new analyzers, this will be merged into main and will be added to 7.0

I've merged from `main`.

----

Spent a bit of time trying to work out how to test adding `System.Linq.Enumerable` overloads but it seems more trouble than it's worth. Anyway, still working on tests, I'll resolve the conversations when ready.

buyaa-n commented:
> After doing this src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.{md,sarif} only contains the first description.

Yes, it is generated with the title/description of each rule, a rule could have multiple messages but should have only one title and one description. Could you integrate the title and description into one? Something like `Do not call Enumerable.Cast or Enumerable.OfType with incompatible types.`

> I've merged from main.

I was wondering why the https://github.com/dotnet/roslyn-analyzers/blob/08382fdbdfe11a6af166d8ebe6110699551f03ab/src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md has only one analyzer while we have added many in 6.0. Got confirmed that `main` is only for nuget releases, no more new analyzers added there, you need to retarget the `release/7.0.1xx` branch, see https://github.com/dotnet/roslyn-analyzers/blob/main/docs/NetCore_GettingStarted.md#branch-definitions for more info

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzer.cs:
```diff
@@ -0,0 +1,319 @@
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotCallEnumerableCastThatWillFailAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string CastRuleId = "CA2017";
+        internal const string OfTypeRuleId = "CA2017";
```
The same id would not work, if you don't want to merge the title and description we can consider using a different id, that might need to go through API review cc @terrajobst

fowl2 commented:
having trouble due to https://github.com/dotnet/docs/issues/24894

thousands of 
> CSC : error AD0001: Analyzer 'Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.PropertiesShouldNotBeWriteOnlyAnalyzer' threw an exception of type 'System.InvalidOperationException' with message 'Passing '.editorconfig' files as addition
> al files is no longer needed. It will be implicitly discovered (if the file is in the project's directory or any ancestor directory), or it should be converted into a 'globalconfig'. See 'https://docs.microsoft.com/dotnet/fundamentals/code
> -analysis/configuration-files'.'. [\roslyn-analyzers\src\Microsoft.CodeAnalysis.AnalyzerUtilities\Microsoft.CodeAnalysis.AnalyzerUtilities.csproj]

Youssef1313 commented:
@fowl2 Can you try deleting this:

https://github.com/dotnet/roslyn-analyzers/blob/1ffb3b40946918f1900d566f2ad3ddd3aaa11bdd/Directory.Build.props#L55-L57


@jeffhandley @buyaa-n Anyone knows how this can happen locally but not in CI? The error seems to make sense to me and it should have been reported by CI. This seems quite odd to me 👀

fowl2 commented:
removing that line seems to have done something

I don't think I've copied the right file, as I'm now getting 3x
> error CA1069: The enum member 'IdeSuggestion' has the same constant value '3' as member 'BuildWarningCandidate'

But nothing for this rule - which should at least be picking up something from its own unit test. 

Which package+version am I supposed to be overwriting? There's about 20 different ones [in here](https://github.com/dotnet/roslyn-analyzers/blob/release/7.0.1xx/eng/Versions.props).

fowl2 commented:
@Youssef1313 do you have any hints as to how to run test against `dotnet/roslyn-analyzers`?

Youssef1313 commented:
> @Youssef1313 do you have any hints as to how to run test against `dotnet/roslyn-analyzers`?

Can you try the instructions [here](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/NetCore_GettingStarted.md#testing-against-the-runtime-and-roslyn-analyzers-repo)?

fowl2 commented:
Yes that's what I was trying. See my previous comment :)

Youssef1313 commented:
Probably @buyaa-n or @mavasani can help better here.

fowl2 commented:
done:
* rebased on main: still builds ✅🎉
* got it running against `roslyn-analyzers`: no problems found
* double checked by adding a problem: found 🎉

to be:
* check against `runtime` (ran into some python dep issue) and `roslyn` (unrelated laptop death)

fowl2 commented:
I've rebased to main, changed to new diagnostic ID (again) and updated to the new `DiagnosticDescriptor`/`LocalizableString` pattern.

Ran out of time/cpu again trying to run it against `runtime` and `rosyln`, could I ask someone else have a go to help get this over the line? :)

buyaa-n reviewed (commented):
> Ran out of time/cpu again trying to run it against runtime and rosyln, could I ask someone else have a go to help get this over the line? :)

Tested with runtime looks good and no diagnostic found, thank you.

Overall looks great to me, left a few questions

fowl2 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs:
```diff
@@ -0,0 +1,722 @@
+        [Fact]
+        public async Task ArrayCSharp()
+        {
+            Assert.Throws<InvalidCastException>(()
+                => (new int[][] { new int[] { 1 } }).Cast<object[]>().ToArray());
+            Assert.Throws<InvalidCastException>(()
+                => (new[] { 1 }).Cast<string>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new object[][] { Array.Empty<object>() }).Cast<int[]>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new ValueType[][] { Array.Empty<ValueType>() }).Cast<int[]>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new object[][] { Array.Empty<ValueType>() }).Cast<int[]>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new ValueType[][] { Array.Empty<ValueType>() }).Cast<int[]>().ToArray());
+
+            Assert.Throws<InvalidCastException>(()
+                => (new int[][] { Array.Empty<int>() }).Cast<ValueType[]>().ToArray());
```
the idea here was two-fold:
1. it was useful to double check the logic while writing the analyzer
2. this could potentially detect if the behavior changes in future runtime/bcl updates. This is pretty unlikely though, so I wouldn't object to removing them.

buyaa-n replied:
I am OK with keeping, defer to @mavasani

fowl2 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs:
```diff
@@ -0,0 +1,722 @@
+        _ = {|#41:(new Apple[0]).Cast<Shoe>()|}; // error
+        _ = {|#42:(new Shoe[0]).Cast<Fruit>()|}; // error
+        _ = {|#43:(new Shoe[0]).Cast<Apple>()|}; // error
+        _ = {|#44:(new Shoe[0]).Cast<Shoe>()|}; // error
```
The comment is wrong - it doesn't warn here and there's no `VerifyCS.Diagnostic(castRule).WithLocation(44)` below.

I've removed the line.

buyaa-n commented:
Thank you @fowl2 LGTM!

@mavasani please look at the open questions and you also have out-dated change request

buyaa-n commented on src/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.md:
```diff
@@ -1860,6 +1860,24 @@ Some built-in operators added in .NET 7 behave differently when overflowing than
+
+The generic type check (C# 'is' operator/IL 'isinst') used by Enumerable.OfType\<T> will never succeed with elements of types specified, resulting in an empty sequence.  
+
+Widening and user defined conversions are not supported with generic types.
```
Confirmed offline that it is OK
